src/renderer/atlas/shader_ps.hlsl

```diff
@@ -183,6 +183,9 @@ float4 main(float4 pos: SV_Position): SV_Target
         [flatten] if (cursorColor == INVALID_COLOR && glyphs[cellPos].a != 0)
         {
             color = float4(1 - color.rgb, 1);
+
+            // Make sure the cursor is always readable (see gh-3647)
+            color = float4((((uint3(color.rgb * 255) ^ 0x3f) & 0xff) / 255.0).rgb, 1);

```

Comment:
I've been thinking about whether this can be expressed in floating point since GPUs prefer floating point over integer operations and because it would allow us to retain the full "color accuracy" up to the end of the shader.

As it turns out, it's possible, but maybe not very readable:
```hlsl
float3 ip; // integral part
// We're trying to map colors 0-255, which is 256 colors and not 255.
// Due to that we multiply by 3.9999999 which causes a `color` of 1
// to map to a `ip` of 3 instead of 4, which prevents the resulting
// color from going negative below.
float3 frac = modf(3.9999999f * color.rgb, ip);
color = float4((3.0f - ip + frac) * 0.25f, 1);
```

I haven't tested this yet, but if you like this please feel free to incorporate it into your PR. 🙂 Otherwise I'll just approve it as it is tomorrow.